app/org/sagebionetworks/bridge/dynamodb/DynamoUserConsentDao.java

```diff
@@ -7,80 +7,106 @@
-import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperConfig.SaveBehavior;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper.FailedBatch;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBQueryExpression;
 import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBScanExpression;
 import com.amazonaws.services.dynamodbv2.model.AttributeValue;
 import com.amazonaws.services.dynamodbv2.model.ComparisonOperator;
 import com.amazonaws.services.dynamodbv2.model.Condition;
-import com.amazonaws.services.dynamodbv2.model.ConditionalCheckFailedException;
 import com.google.common.collect.Sets;
 
 @Component
 public class DynamoUserConsentDao implements UserConsentDao {
 
-    private DynamoDBMapper mapper;
+    private DynamoDBMapper mapperV2;
+    private DynamoDBMapper mapperV3;
+
+    @Resource(name = "userConsentDdbMapper2")
+    public final void setDdbMapper2(DynamoDBMapper mapper2) {
+        this.mapperV2 = mapper2;
+    }
 
-    @Autowired
-    public void setDynamoDbClient(BridgeConfig bridgeConfig, AmazonDynamoDB client) {
-        DynamoDBMapperConfig mapperConfig = new DynamoDBMapperConfig.Builder().withSaveBehavior(SaveBehavior.UPDATE)
-                .withConsistentReads(ConsistentReads.CONSISTENT)
-                .withTableNameOverride(DynamoUtils.getTableNameOverride(DynamoUserConsent2.class, bridgeConfig)).build();
-        mapper = new DynamoDBMapper(client, mapperConfig);
+    @Resource(name = "userConsentDdbMapper3")
+    public final void setDdbMapper3(DynamoDBMapper mapper3) {
+        this.mapperV3 = mapper3;
     }
 
     @Override
     public UserConsent giveConsent(String healthCode, StudyConsent studyConsent) {
         checkArgument(isNotBlank(healthCode), "Health code is blank or null");
         checkNotNull(studyConsent);
-        DynamoUserConsent2 consent = null;
-        try {
-            consent = getUserConsent(healthCode, studyConsent);
-            if (consent == null) {
-                consent = new DynamoUserConsent2(healthCode, studyConsent);
-            }
-            consent.setSignedOn(DateTime.now(DateTimeZone.UTC).getMillis());
-            mapper.save(consent);
-        } catch (ConditionalCheckFailedException e) {
-            throw new EntityAlreadyExistsException(consent);
+
+        long signedOn = DateTime.now(DateTimeZone.UTC).getMillis();
+
+        // The logic is being changed here. You cannot update the signing date on an active
+        // consent record if it exists. Throw an exception if it does.
+        DynamoUserConsent2 consent2 = getUserConsent2(healthCode, studyConsent.getStudyKey());
+        if (consent2 != null) {
+            throw new EntityAlreadyExistsException(consent2);
+        }
+        
+        // save 3 first because 2 is the active record. If 3 fails that's not a problem and can 
+        // be fixed in migration, but if 2 fails, consent didn't occur.
```

Comment:
I think you have this backwards. As written, if 3 fails, 2 will never be called, so consent will never be recorded. If 3 succeeds and 2 fails, we have a phantom consent record.

Also, even if we flip the order, if 2 succeeds and 3 fails, we return an error to the user. I think we need a try-catch around consent3 and log an error, but recover.